flowmachine/flowmachine/core/cache.py

```diff
@@ -0,0 +1,246 @@
+"""
+import logging
+import pickle
+
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from .query import Query
+    from .connection import Connection
+
+logger = logging.getLogger("flowmachine").getChild(__name__)
+
+
+def get_query_by_id(connection: "Connection", query_id: str) -> "Query":
+    """
+    Get a query object from cache by id.
+
+    Parameters
+    ----------
+    connection : Connection
+    query_id : str
+        md5 id of the query
+
+    Returns
+    -------
+    Query
+        The original query object.
+
+    """
+    qry = f"SELECT obj FROM cache.cached WHERE query_id='{query_id}'"
+    try:
+        obj = connection.fetch(qry)[0][0]
+        return pickle.loads(obj)
+    except IndexError:
+        raise ValueError(f"Query id '{query_id}' is not in cache on this connection.")
+
+
+def shrink_one(connection: "Connection", dry_run: bool = False) -> "Query":
+    """
+    Remove the lowest scoring cached query from cache and return it and size of it
+    in bytes.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    dry_run : bool, default False
+        Set to true to just report the object that would be removed and not remove it
+
+    Returns
+    -------
+    "Query"
+        The "Query" object that was removed from cache
+    """
+    qry = "SELECT tablename, schema, obj FROM cache.cached WHERE NOT class='Table' ORDER BY cache_score ASC LIMIT 1"
+    tablename, schema, obj = connection.fetch(qry)[0]
+    table_size = size_of_table(connection, tablename, schema)
+    obj_to_remove = pickle.loads(obj)
+
+    logger.info(
+        f"{'Would' if dry_run else 'Will'} remove cache record for {obj_to_remove.md5} of type {obj_to_remove.__class__}"
+    )
+    logger.info(
+        f"Table {schema}.{tablename} ({table_size} bytes) {'would' if dry_run else 'will'} be removed."
+    )
+
+    if not dry_run:
+        obj_to_remove.invalidate_db_cache(
+            name=tablename, schema=schema, cascade=False, drop=True
+        )
+    return obj_to_remove, table_size
+
+
+def shrink_below_size(
+    connection: "Connection", size_threshold: int, dry_run: bool = False
+) -> "Query":
+    """
+    Remove queries from the cache until it is below a specified size threshold.
+
+    Parameters
+    ----------
+    connection : "Connection"
+    size_threshold : int
+        Size (in bytes) to reduce the cache below
+    dry_run : bool, default False
+        Set to true to just report the objects that would be removed and not remove them
+
+    Returns
+    -------
+    list of "Query"
+        List of the queries that were removed
+    """
+    initial_cache_size = size_of_cache(connection)
+    removed = []
+    logger.info(
+        f"Shrinking cache from {initial_cache_size} to below {size_threshold}{'(dry run)' if dry_run else ''}."
+    )
+
+    while initial_cache_size > size_threshold:
+        obj_removed, cache_reduction = shrink_one(connection, dry_run)
```

Comment:
Will this work in dry-run if more than one query would be removed? Since `shrink_one` won't actually remove `obj_removed` in a dry run, it looks like the same query would be picked up again in the next iteration of the loop.

Reply:
Good point!